### ISSUE TITLE
Missing GPU op for zeros_like for Ragged Tensor, error occurs when Ragged Tensor fed thru tf.map_fn

### ISSUE THREAD

**Issue by djoshea** [tensorflow/tensorflow#46635]
**System information**

Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes, included below
OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 20.10
Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: n/a
TensorFlow installed from (source or binary): pip binary
TensorFlow version (use command below): v1.12.1-49539-g18d8bcbe72b 2.5.0-dev20210123
Python version: '3.8.6 | packaged by conda-forge | (default, Nov 27 2020, 19:31:52) \n[GCC 9.3.0]'
Bazel version (if compiling from source): n/a
GCC/Compiler version (if compiling from source): n/a
CUDA/cuDNN version: 11.0 / 8
GPU model and memory: TITAN X (Pascal) computeCapability: 6.1

**Describe the current behavior**

I have a keras layer `RescaleB` that accepts a ragged tensor with shape [batch, (time), in_dim]. The layer calls `map_fn` to process each example in the batch separately, scaling the values along the inner dimension by a trainable gain vector. (The details of the operation aren't critical, but the ragged tensor going into map_fn is.)

Using this layer fails with `No unary variant unary_op function found for unary variant op enum: 1 Variant type_name: RaggedTensorVariant for device type: GPU` on a node whose name ends with `rescale_b/map/while/TensorArrayV2Write/TensorListSetItem_grad/zeros_like` which suggests that the zeros_like operation 
isn't defined for Ragged Tensors on GPU?

In this simple example, i also include `RescaleA`, which accomplishes the same task using `tf.ragged.map_flat_values`, although in my real use case I need `map_fn`. This is a simplified example.

**Describe the expected behavior**

I'd expect `RescaleB` and `RescaleA` to function identically.

**Standalone code to reproduce the issue**

https://colab.research.google.com/drive/1mHycCXJL94VuCGkXIJ0bIXtbYamyZo78

I've reproduced the issue locally with tf-nightly-gpu TF 2.5, but I can't seem to get the nightly version to see the GPU on Colab. The Colab notebook is using TF 2.4, but the issue remains in TF 2.5 nightly.

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

This may be the same issue as #44231 but hopefully the additional detail here is helpful.

**Comment by Saduf2019** [comment]
I ran the code on tf 2.4 and and face a different error on nightly, please find the [gist here](https://colab.research.google.com/gist/Saduf2019/e6435e077a9cd287a043bed1389da761/untitled507.ipynb)

**Comment by djoshea** [comment]
As I mentioned, I don't think tf-nightly-gpu is running on the GPU in Colab. The issue is only present on GPU. In the gist you sent, the cell where it searches for the GPU returns:
```
Tensorflow version ==  2.5.0-dev20210124

---------------------------------------------------------------------------
SystemError                               Traceback (most recent call last)
<ipython-input-8-6f54130e32b7> in <module>()
      4 print(device_name)
      5 if device_name != '/device:GPU:0':
----> 6   raise SystemError('GPU device not found')
      7 print('Found GPU at: {}'.format(device_name))

SystemError: GPU device not found
```

Locally, where it is running on GPU, TF nightly fails with the same error as TF2.4, so this issue is still present in nightly, at least on my local `pip install tf-nightly-gpu` installation.

**Comment by sanjoy** [comment]
@djoshea this looks like a reasonable request, will you be able to send a PR for this?

**Comment by djoshea** [comment]
I could potentially try developing it, but I unfortunately don't know where to start. Is there a guide somewhere to implementing new ops for the GPU?

**Comment by sanjoy** [comment]
> I could potentially try developing it, but I unfortunately don't know where to start. Is there a guide somewhere to implementing new ops for the GPU?

I don't think we have a detailed guide for this, but if you're comfortable with C++ you could try to start [at the check that's failing](https://github.com/tensorflow/tensorflow/blob/5ecbbd87a4556e06cbb3dcf384d088dd27fce4e0/tensorflow/core/framework/variant_op_registry.h#L317).

**Comment by patchmeifyoucan** [comment]
It would be nice if this could be implemented soon. Using map_fn with RaggedTensors is very convenient when working with data of different shapes. Unfortunately one must run these operations on CPU right now.

**Comment by areiner222** [comment]
I'm relying on map_fn to extract ragged image patches using bounding boxes so I can then produce ragged bounding-box centered feature maps via convolutions. I am encountering the same bug with the gradient step inside a tf.function (I can successfully run outside tf.function). The ragged image patches method is enabling a large speed up with a reduced memory footprint so it would be very useful to enable this in graph mode on GPUs.

**Comment by patchmeifyoucan** [comment]
For me RaggedTensors + map_fn are quite an enabler because I am implementing some kind of link prediction in a graph neural network. Having TensorFlow handling the individual samples allows for pretty safe programming as I do not have to put all samples in a batch into a large, disconnected graph and make sure that no nodes of different graphs will be connected. Using TensorFlow this way results in nice code that is fast to write in contrast to solutions with masks which may need to be changed every time you change your computation. Having this in GPU would be awesome. :100:

**Comment by OliverGuy** [comment]
I would love to see this fixed, especially as iterating over RaggedTensors seems to not properly work in a lot of cases.

**Comment by sushreebarsa** [comment]
I ran the code in TF v2.5 and face error ,please find the gist [here](https://colab.research.google.com/gist/sushreebarsa/0bf4ae93db25542263e443bad439fb94/untitled155.ipynb?authuser=1)..Thanks !

**Comment by llan-ml** [comment]
Hi @sanjoy any updates on this? This is a really needed feature.

**Comment by timolange** [comment]
Hi, i will also very appreciate if this bug get fixed.
I think it's somehow related to [No gradient defined for operation RaggedTensorFromVariant](https://github.com/tensorflow/tensorflow/issues/42189) and [No gradient defined for operation RaggedTensorFromVariant / or no gradients at all
](https://github.com/tensorflow/tensorflow/issues/43107) which are fixed with commit [be6b1fd](https://github.com/tensorflow/tensorflow/commit/be6b1fdb0699d4000b70ad32cc23d1503e5c7511) by @edloper .

The part with `REGISTER_UNARY_VARIANT_UNARY_OP_FUNCTION(
    ZEROS_LIKE_VARIANT_UNARY_OP, DEVICE_CPU, RaggedTensorVariant,
    RaggedTensorVariantZerosLike<CPUDevice>);` just registers the zeros_like OP for CPU device but not for GPU.

Maybe this will be a quick win to fix.
@edloper ,it would be very kind, if you can have a look at this.
Thanks.

**Comment by JXRiver** [comment]
Hi, thanks for raising the issue. According to @edloper "Basically, RaggedTensorVariant objects should never be copied to GPU, because we can't do anything useful with them there. But Placer isn't currently smart enough to figure that out (it just sees a Variant tensor, and doesn't know what kind of value it contains)." We have a project going on right now that hopefully will fix the issue.

**Comment by Marcus2810** [comment]
Okay, I know this may sound too easy but, try using: to_tensor() on the ragged_tensor and continue from there.
But say per batch is a different shape: then np.array([ TF.MAPFN_HERE for batch in batches]), but make sure the value after mapped by list compression is changed from raggedTensor to tensor.

Hope this helps.

**Comment by misha-antonenko** [comment]
Hello. Could you please inform if there has been any progress on the issue?

For me, it resulted in a sudden sixfold increase of the train step length due to having to move to CPU

**Comment by ablanch5** [comment]
Also needing this resolved

**Comment by isarandi** [comment]
I faced the same issue and resolved it by rolling my own `map_fn` via `TensorArray`. Here's a rough sketch:

```python
def map_fn2(fn, elems, fn_output_signature):
    arr = tf.TensorArray(fn_output_signature.dtype, size=fn_output_signature.shape[0])
    batch_size = tf.nest.flatten(elems)[0].shape[0]
    for i in tf.range(batch_size):
        arr = arr.write(i, fn(tf.nest.map_structure(lambda x: x[i], elems)))
    return arr.stack()
```

**Comment by ivkhar** [comment]
As of TF v. 2.15, still not resolved.
It would be nice if the limitation on using RaggedTensors with map_fn on GPU will be mentioned on map_fn documentation.

**Comment by Ryandry1st** [comment]
Agreed, I would like to see this resolved as it is pretty important functionality for using map_fn and ragged tensors.

**Comment by forest-spells** [comment]
I just wanted to let you know, as for 2025, people still have hope for resolving this

**Comment by Venkat6871** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
